### PR TITLE
Add nubOrd and nubOrdBy

### PR DESCRIPTION
**Pull request by Boarders**
This adds nubOrd functions addressing https://github.com/haskell/vector/issues/145. The implementation of a red black tree for testing membership is taken directly from the `extra` package.

I have not looked at the internals of vector before seriously so apologies for anything glaringly wrong.

The issue also mentions adding uniqBy which looks completely straightforward and I am happy to do in a separate PR.

### CODE REVIEW

File: Data/Vector/Fusion/Stream/Monadic.hs [hunk 1]
```diff
@@ -794,6 +795,67 @@ dropWhileM f (Stream step t) = Stream step' (DropWhile_Drop t)
...
+nubOrdBy :: Monad m => (a -> a -> Ordering) -> Stream m a -> Stream m a
+{-# INLINE_FUSED nubOrdBy #-}
+nubOrdBy cmp (Stream step st) = Stream step' (E, st)
+  where
+    {-# INLINE_INNER step' #-}
+    step' (set, s)
+      = do
+          r <- step s
+          case r of
+            Yield x s' | memberRB cmp x set -> return $ Skip (insertRB cmp x set, s')
```
[the remark this answers is not in the document]
Comment (Boarders) [reply Data/Vector/Fusion/Stream/Monadic.hs:814]: Oh how silly, let me fix that.

File: Data/Vector/Fusion/Stream/Monadic.hs [hunk 2]
```diff
@@ -794,6 +795,67 @@ dropWhileM f (Stream step t) = Stream step' (DropWhile_Drop t)
...
+            Yield x s' | otherwise           -> return $ Yield x (insertRB cmp x set, s')
```
[the remark this answers is not in the document]
Comment (Boarders) [reply Data/Vector/Fusion/Stream/Monadic.hs]: Now fixed.

**Comment by Boarders** [comment]
I added some documentation for what is expected to get reasonable behaviour. I will try to add some benchmarks later. I may need some suggestions as the current benchmarks simply run some random algorithms. I could compare to the implementation for list or I could run for different sizes to make sure it has the correct asymptotic. What do you think? The benchmarks don't do anything for memory so I am unsure if there is a good approach for that.

**Comment by Boarders** [comment]
I cannot build the benchmarks with cabal currently, what is the intended way to build?

**Comment by Boarders** [comment]
I needed to add the following to the `cabal.project` file in benchmarks in order for it to build:
```
allow-newer:
  vector-algorithms:vector
  vector-th-unbox:vector
  microstache:vector
  cassava:vector
  aeson:vector
```

It might be worth coming up with a good way to deal with it.

**Comment by Shimuuar** [comment]
I just tried to suggest adding allow-newer to cabal.project file. Moving to gauge could be an option. AFAIU it's criterion with dependencies partially removed and partially vendored.

WRT to suggestions for benchmarks I have many ideas! Maybe too many. We have many variables:

 1. Length of vector (obviously)
 2. Cost of comparison. Cheap (e.g. `Int`) will be sensitive to overhead of RB-tree. Expensive (longish `String` with common prefix) will be affected mostly by number of comparisons
 3. Fraction of duplicates which affects growth speed of tree
 4. Whether elements ordered/nearly ordered/random. This affects how tree is going to be rebalaced. If algorithm has bug it may cost us dearly

What we need is to check that we perform comparable with `Data.List.nub` for short lists to check that we don't have bad constant factors, vary length to check asymptotics, and it would be fine to check that being ordered/disordered doesn't break performance. Everything else is extra

**Comment by Boarders** [comment]
Current implementation is ridiculously slow so I will have to investigate what is going on.

**Comment by Boarders** [comment]
At the moment this implementation is embarrassingly slow:
```
benchmarking nubOrd
time                 10.65 ns   (10.40 ns .. 10.95 ns)
                     0.995 R²   (0.991 R² .. 0.998 R²)
mean                 10.48 ns   (10.30 ns .. 10.76 ns)
std dev              714.3 ps   (492.1 ps .. 1.012 ns)
variance introduced by outliers: 84% (severely inflated)

benchmarking nubOrd
time                 6.703 s    (5.783 s .. 7.657 s)
                     0.997 R²   (0.990 R² .. 1.000 R²)
mean                 7.045 s    (6.813 s .. 7.259 s)
std dev              269.3 ms   (127.2 ms .. 341.6 ms)
variance introduced by outliers: 19% (moderately inflated)
```

I don't know if this is because carrying around large state in stream fusion is a bad idea or something else? If anyone has any ideas let me know. I also tried using the implementation of `Data.Set` for insertion and membership but it didn't make a difference so the problem must be something more fundamental.

**Comment by Bodigrim** [comment]
To be honest, I'm not completely on board with RB-tree here. In `vector` I expect deterministic and predictable allocations. Imagine calling `nubOrd` on an unboxed vector - RB-tree will easily eat 10x-20x more memory and cause a significant pressure on GC. 

Would not it be better just `thaw` the original vector, sort it in-place, removing duplicates, followed by `slice` and `unsafeFreeze`? Even if it means sacrificing fusion.

**Comment by Boarders** [comment]
I can give that a try (though I would be interested why the current version has such awful performance). I don't easily see how you do this in a stateless way, if you could elaborate then I could try it.

**Comment by chessai** [comment]
I agree with Bodigrim here based on my intuition. I would expect an in-place sort on a mutable vector, then a duplicates removal pass on the sorted array. Sorting is much more expensive than duplicate removal. See [primitive-sort](https://hackage.haskell.org/package/primitive-sort-0.1.0.0/docs/Data-Primitive-Sort.html)

**Comment by chessai** [comment]
Another idea is to use a heap sort with heap insert modified to remove duplicate elements. worse cache locality, but should be fast and with few allocations. Might be worthwhile to investigate and benchmark.

**Comment by Bodigrim** [comment]
Wouldn't it be worth to retarget this PR to `vector-algorithms`? It is an appropriate place for all kinds of sorts and similar stuff.

**Comment by Bodigrim** [comment]
```haskell
benchmarking List.nubOrd
time                 10.65 ns   (10.40 ns .. 10.95 ns)
```

This looks suspiciously fast, 10ns is just several dozens of CPU instructions. Could you possibly check that your benchmark forces evaluation of the full list and not only its head?

**Comment by Boarders** [comment]
@Bodigrim Ah ok, let me investigate because it seemed quite ridiculous. I think you are correct.

@Bodigrim @chessai: what you are describing sounds like `sort . nubOrd` and not `nubOrd`. I do not think we should give such a function the same name as that would be very confusing. I will investigate your heap sort idea though.

**Comment by Bodigrim** [comment]
@Boarders You can implement an order-preserving `nubOrd` atop of `sortUnique`.

Once you have sorted and deduplicated vector, add a mutable bit mask to keep track of used elements. Now traverse through the original vector once again, for each element find its position in the sorted vector (via binary search) and check the correspondent bit of the bit mask. If it is 0, flip it to 1 and write the element into a return vector. If it is 1, do nothing. 

There may be more elegant ways to do this, using less additional vectors, but decoupling from the choice of sorting routine sounds desirable to me.

**Comment by Boarders** [comment]
@Bodigrim The above was indeed a false alarm and the performance is better than (but comparable to) the version from `extra`. Thank you for pointing it out, I thought I was going mad.

**Comment by Boarders** [comment]
Just for those interested, using the relevant bits from `Data.Set` leads to a good speed up here though it looks like using any data structure here is unwanted. In practice I am not sure how much of an issue it would be but I am happy for this to be closed as the approach will not work.

**Comment by Bodigrim** [comment]
@Boarders I do appreciate your efforts and I believe `nubOrd` would make a suitable contribution to `vector-algorithms`. 

As explained above, using trees to `nubOrd` an unboxed vector of `Word8` will lead to intermediate memory consumption magnitudes larger than the input vector itself. I think that such algorithms are too high level and opinionated for `vector`, and thus out of scope.

**Comment by Shimuuar** [comment]
Yes. nub & friends are tightly coupled with sorting algorithms. So proper solution will probably require tight integration with the `vector-algorithms`. 

However sorting of vectors is sort of joke too. `vector-algorithms` only provides API for mutable vectors. No default. Which algorithm should who know very little about sorting pick?

**Comment by Bodigrim** [comment]
@Shimuuar Let's discuss the lack of `sort` in #232

**Comment by Boarders** [comment]
@Bodigrim : No problem, when I get the chance I will investigate how to get a reasonable memory version of nubOrd using a sort and mask or other technique and where to put that.

**Comment by Shimuuar** [comment]
So approach with RB trees didn't work out. So I'm closing this PR. Reopen if you disagree.

**Comment by Boarders** [comment]
@Bodigrim I implemented your suggestion here btw: https://github.com/erikd/vector-algorithms/pull/35

**Pull request closed without merging**